File: 2022/04.mjs
import { readAndSolve } from '../aoc.mjs'

const lineToSections = (line) => line.split(/[,-]/).map(x => +x).chunk(2);

const isWhollyContainedIn = ([left, right]) =>
    (left[0] >= right[0] && left[1] <= right[1])
        || (right[0] >= left[0] && right[1] <= left[1]);

const isPartiallyContainedIn = ([left, right]) => 
         (left[0] >= right[0] && left[0] <= right[1])
            || (left[1] >= right[0] && left[1] <= right[1])
            || (right[0] >= left[0] && right[0] <= left[1])
            || (right[1] >= left[0] && right[1] <= left[1]);

function solve(lines) {
    const p1 = lines.map(lineToSections).map(isWhollyContainedIn).filter( x => x).length;
    const p2 = lines.map(lineToSections).map(isPartiallyContainedIn).filter( x => x).length;

    return {p1, p2};
};

readAndSolve(process.argv[2] || '04.input', solve);

File: 2022/04.exs
Code.require_file("../aoc.exs")

defmodule Puzzle do
    def solve(instructions) do
        solver(instructions, &isWhollyContained/2)
    end

    def solve2(instructions) do
        solver(instructions, &isPartiallyContained/2)
    end

    defp solver(instructions, check) do
        instructions
        |> Enum.map(fn pair ->
            [left, right] = String.split(pair, ",")
            left = String.split(left, "-") |> Enum.map(&String.to_integer/1)
            right = String.split(right, "-") |> Enum.map(&String.to_integer/1)
            check.(left, right)
        end)
        |> Enum.count(fn x -> x end)
    end

    def isWhollyContained(left, right) do
        [l0, l1] = left
        [r0, r1] = right
        (l0 >= r0 && l1 <= r1) || (r0 >= l0 && r1 <= l1);
    end

    def isPartiallyContained(left, right) do
        [l0, l1] = left
        [r0, r1] = right
         (l0 >= r0 && l0 <= r1)
            || (l1 >= r0 && l1 <= r1)
            || (r0 >= l0 && r0 <= l1)
            || (r1 >= l0 && r1 <= l1)
    end

end

Aoc.readAndSolve("04.input", &Puzzle.solve/1)
|> IO.inspect(label: "Part 1")

Aoc.readAndSolve("04.input", &Puzzle.solve2/1)
|> IO.inspect(label: "Part 2")
